lib/std/crypto/25519/curve25519.zig

```diff
@@ -0,0 +1,155 @@
+const std = @import("std");
+
+/// Group operations over Curve25519.
+pub const Curve25519 = struct {
+    /// The underlying prime field.
+    pub const Fe = @import("field25519.zig").Fe;
+    /// Field arithmetic mod the order of the main subgroup.
+    pub const scalar = @import("scalar25519.zig");
+
+    x: Fe,
+
+    /// Decode a Curve25519 point from its compressed (X) coordinates.
+    pub inline fn fromBytes(s: [32]u8) Curve25519 {
+        return .{ .x = Fe.fromBytes(s) };
+    }
+
+    /// Encode a Curve25519 point.
+    pub inline fn toBytes(p: Curve25519) [32]u8 {
+        return p.x.toBytes();
+    }
+
+    /// Return the Curve25519 base point.
+    pub inline fn basePoint() Curve25519 {
+        return .{ .x = Fe.curve25519BasePoint() };
+    }
+
+    /// Check that the encoding of a Curve25519 point is canonical.
+    pub fn rejectNonCanonical(s: [32]u8) !void {
+        return Fe.rejectNonCanonical(s, false);
+    }
+
+    /// Reject the neutral element.
+    pub fn rejectIdentity(p: Curve25519) !void {
+        if (p.x.isZero()) {
+            return error.IdentityElement;
+        }
+    }
+
+    fn ladder(p: Curve25519, s: [32]u8, comptime bits: usize) !Curve25519 {
+        var x1 = p.x;
+        var x2 = Fe.one();
+        var z2 = Fe.zero();
+        var x3 = x1;
+        var z3 = Fe.one();
+        var swap: u8 = 0;
+        var pos: usize = bits - 1;
+        while (true) {
+            const b = (s[pos / 8] >> @intCast(u3, pos & 7)) & 1;
+            swap ^= b;
+            Fe.cSwap2(&x2, &x3, &z2, &z3, swap);
+            swap = b;
+            var tmp0 = x3.sub(z3);
+            var tmp1 = x2.sub(z2);
+            x2 = x2.add(z2);
+            z2 = x3.add(z3);
+            z3 = tmp0.mul(x2);
+            z2 = z2.mul(tmp1);
+            tmp0 = tmp1.sq();
+            tmp1 = x2.sq();
+            x3 = z3.add(z2);
+            z2 = z3.sub(z2);
+            x2 = tmp1.mul(tmp0);
+            tmp1 = tmp1.sub(tmp0);
+            z2 = z2.sq();
+            z3 = tmp1.mul32(121666);
+            x3 = x3.sq();
+            tmp0 = tmp0.add(z3);
+            z3 = x1.mul(z2);
+            z2 = tmp1.mul(tmp0);
+            if (pos == 0) break;
+            pos -= 1;
+        }
+        Fe.cSwap2(&x2, &x3, &z2, &z3, swap);
+        z2 = z2.invert();
+        x2 = x2.mul(z2);
+        if (x2.isZero()) {
+            return error.IdentityElement;
```

Comment:
Do you think error values introduced in this PR would benefit from enclosing inside an explicit error enum set for re-use in different curve/crypto schemes implementations that might arise in the future?

Reply:
Yes, and this is an excellent idea!

Other primitives can also benefit from reusable error sets. This is something I'd be happy to work on.

Reply:
Can this be addressed in a distinct PR?

Reply:
Sounds good to me!